pkg/reconciler/buildrun/runner/taskrun.go

```diff
@@ -0,0 +1,491 @@
+	payload := []patchStringValue{{
+		Op:    op,
+		Path:  path,
+		Value: value,
+	}}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+	patch := client.RawPatch(types.JSONPatchType, data)
+	patchOpt := client.PatchOptions{Raw: &opts}
+	return k8sClient.Patch(ctx, tr, patch, &patchOpt)
+}
+
+func (t *TaskRunBuildRunner) SyncBuildRunStatus(ctx context.Context, client client.Client, buildRun *v1beta1.BuildRun) error {
+
+	if buildRun.Status.StartTime == nil && t.GetStartTime() != nil {
+		buildRun.Status.StartTime = t.GetStartTime()
+	}
+
+	if buildRun.Status.CompletionTime == nil && t.GetCompletionTime() != nil {
+		buildRun.Status.CompletionTime = t.GetCompletionTime()
+	}
+
+	if len(t.Status.Results) > 0 {
+		// TODO: this logic could be optimized so we iterate through the result values only once.
+		updateBuildRunStatusWithSourceResult(buildRun, t.Status.Results)
+		updateBuildRunStatusWithOutputResult(buildRun, t.Status.Results)
+	}
+
+	succeededCondition := t.Status.GetCondition(apis.ConditionSucceeded)
+	if succeededCondition == nil {
+		return nil
+	}
+
+	if err := t.updateBuildRunUsingTaskRunCondition(ctx, client, buildRun, succeededCondition); err != nil {
+		return err
+	}
+
+	t.updateBuildRunUsingTaskFailures(ctx, client, buildRun)
+
+	return nil
+}
+
+const (
+	defaultSourceName          = "default"
+	sourceTimestampName        = "source-timestamp"
+	prefixParamsResultsVolumes = "shp"
+	imageDigestResult          = "image-digest"
+	imageSizeResult            = "image-size"
+	imageVulnerabilities       = "image-vulnerabilities"
+)
+
+func updateBuildRunStatusWithSourceResult(buildrun *v1beta1.BuildRun, results []pipelineapi.TaskRunResult) {
+	buildSpec := buildrun.Status.BuildSpec
+
+	if buildSpec.Source == nil {
+		return
+	}
+
+	switch {
+	case buildSpec.Source.Type == v1beta1.OCIArtifactType && buildSpec.Source.OCIArtifact != nil:
+		sources.AppendBundleResult(buildrun, defaultSourceName, results)
+
+	case buildSpec.Source.Type == v1beta1.GitType && buildSpec.Source.Git != nil:
+		sources.AppendGitResult(buildrun, defaultSourceName, results)
+	}
+
+	if sourceTimestamp := sources.FindResultValue(results, defaultSourceName, sourceTimestampName); strings.TrimSpace(sourceTimestamp) != "" {
+		if sec, err := strconv.ParseInt(sourceTimestamp, 10, 64); err == nil {
+			if buildrun.Status.Source != nil {
+				buildrun.Status.Source.Timestamp = &metav1.Time{Time: time.Unix(sec, 0)}
+			}
+		}
+	}
+}
+
+func updateBuildRunStatusWithOutputResult(buildRun *v1beta1.BuildRun, taskRunResult []pipelineapi.TaskRunResult) {
+	if buildRun.Status.Output == nil {
+		buildRun.Status.Output = &v1beta1.Output{}
+	}
+
+	for _, result := range taskRunResult {
+		switch result.Name {
+		case generateOutputResultName(imageDigestResult):
+			buildRun.Status.Output.Digest = result.Value.StringVal
+
+		case generateOutputResultName(imageSizeResult):
+			// TODO: Throw error that should not be retried?
```

Comment:
I suggest to log a warning if the conversion here fails. Usually means that the build strategy is incorrect in the case of strategy-managed push.